File: AccuPay/Domain/Shifts/EmployeeDutySchedule.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema

Namespace Global.AccuPay.Entity

    <Table("shiftschedules")>
    Public Class EmployeeDutySchedule

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer

        Public Property OrganizationID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As DateTime

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property LastUpd As DateTime

        Public Property LastUpdBy As Integer?

        <ForeignKey("Employee")>
        Public Property EmployeeID As Integer?

        <Column("Date")>
        Public Property DateSched As Date

        Public Property StartTime As TimeSpan?
        Public Property EndTime As TimeSpan?
        Public Property BreakStartTime As TimeSpan?
        Public Property BreakLength As Decimal
        Public Property IsRestDay As Boolean
        Public Property ShiftHours As Decimal
        Public Property WorkHours As Decimal

        Public Overridable Property Employee As Employee

    End Class

End Namespace

File: Accupay.Data/Entities/EmployeeDutySchedule.cs
using AccuPay.Data.Helpers;
using AccuPay.Utilities.Extensions;
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("shiftschedules")]
    public class EmployeeDutySchedule
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int RowID { get; set; }

        public int? OrganizationID { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime Created { get; set; }

        public int? CreatedBy { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime LastUpd { get; set; }

        public int? LastUpdBy { get; set; }

        [ForeignKey("Employee")]
        public int? EmployeeID { get; set; }

        [Column("Date")]
        public DateTime DateSched { get; set; }

        public TimeSpan? StartTime { get; set; }
        public TimeSpan? EndTime { get; set; }
        public TimeSpan? BreakStartTime { get; set; }
        public decimal BreakLength { get; set; }
        public bool IsRestDay { get; set; }
        public decimal ShiftHours { get; internal set; }
        public decimal WorkHours { get; internal set; }

        public virtual Employee Employee { get; set; }

        [NotMapped]
        public DateTime? ShiftStartTimeFull
        {
            get => StartTime == null ?
                        (DateTime?)null :
                        DateSched.Date.ToMinimumHourValue().Add(StartTime.Value);

            set => StartTime = value == null ? null : value?.TimeOfDay;
        }

        [NotMapped]
        public DateTime? ShiftEndTimeFull
        {
            get => EndTime == null ?
                        (DateTime?)null :
                        DateSched.Date.ToMinimumHourValue().Add(EndTime.Value);

            set => EndTime = value == null ? null : value?.TimeOfDay;
        }

        [NotMapped]
        public DateTime? ShiftBreakStartTimeFull
        {
            get => BreakStartTime == null ?
                        (DateTime?)null :
                        DateSched.Date.ToMinimumHourValue().Add(BreakStartTime.Value);

            set => BreakStartTime = value == null ? null : value?.TimeOfDay;
        }

        /// <summary>
        /// Computes the shift hours and also update the work hours.
        /// </summary>
        public void ComputeShiftHours()
        {
            if (StartTime.HasValue && EndTime.HasValue)
            {
                var trueEndTime = StartTime > EndTime ? EndTime.Value.AddOneDay() : EndTime.Value;

                double totalMinutes = (trueEndTime - StartTime).Value.TotalMinutes;

                ShiftHours = Convert.ToDecimal(totalMinutes / TimeConstants.MinutesPerHour);
            }
            else
            {
                ShiftHours = 0;
            }

            ComputeWorkHours();
        }

        private void ComputeWorkHours() => WorkHours = ShiftHours > BreakLength ? ShiftHours - BreakLength : 0;
    }
}